src/user/MOM_wave_interface.F90

```diff
@@ -1876,30 +1946,51 @@ subroutine ust_2_u10_coare3p5(USTair, U10, GV, US, CS)
     "ust_2_u10_coare3p5 called with a negative value of Waves%vonKar")
 
   z0sm = 0.11 * CS%nu_air / USTair ! Compute z0smooth from ustar guess
-  u10 = US%Z_to_L*USTair / sqrt(0.001)  ! Guess for u10
-  !### For efficiency change the line above to USTair * sqrt(1000.0) or USTair * 31.6227766 .
   u10a = 1000.0*US%m_s_to_L_T ! An insanely large upper bound for u10.
 
-  CT=0
-  do while (abs(u10a/u10 - 1.) > 0.001)  !### Change this to (abs(u10a - u10) > 0.001*u10) for efficiency.
-    CT=CT+1
-    u10a = u10
-    alpha = min(0.028, 0.0017*US%L_T_to_m_s * u10 - 0.005)
-    z0rough = alpha * (US%Z_to_L*USTair)**2 / GV%g_Earth ! Compute z0rough from ustar guess
-    z0 = z0sm + z0rough
-    CD = ( CS%vonKar / log(10.*US%m_to_Z / z0) )**2 ! Compute CD from derived roughness
-    u10 = US%Z_to_L*USTair/sqrt(CD)  ! Compute new u10 from derived CD, while loop
-                           ! ends and checks for convergence...CT counter
-                           ! makes sure loop doesn't run away if function
-                           ! doesn't converge.  This code was produced offline
-                           ! and converged rapidly (e.g. 2 cycles)
-                           ! for ustar=0.0001:0.0001:10.
-    if (CT>20) then
-      u10 = US%Z_to_L*USTair/sqrt(0.0015) ! I don't expect to get here, but just
-                              !  in case it will output a reasonable value.
-      exit
-    endif
-  enddo
+  if (CS%answer_date < 20230103) then
+    u10 = US%Z_to_L*USTair / sqrt(0.001)  ! Guess for u10
+    ten_m_scale = 10.0*US%m_to_Z
+    CT=0
+    do while (abs(u10a/u10 - 1.) > 0.001)
+      CT=CT+1
+      u10a = u10
+      alpha = min(CS%Charnock_min, CS%Charnock_slope_U10 * u10 + CS%Charnock_intercept)
+      z0rough = alpha * (US%Z_to_L*USTair)**2 / GV%g_Earth ! Compute z0rough from ustar guess
+      z0 = z0sm + z0rough
+      Cd2 = ( CS%vonKar / log(ten_m_scale / z0) )**2 ! Compute CD from derived roughness
```

Comment:
This is the expression for what is usually given as Cd, not Cd^2, since Cd is the factor between wind squared and wind stress.  In the following calculation it is thus the square root of Cd that relates wind to ustar.  So it is probably clearer to call this "Cd" here (The new improvement below is absolutely correct though by not squaring the expression since the square root of Cd is what we need!).

Reply:
Thanks for this clarification. I have updated the commits to reflect these comments.